### ISSUE TITLE
Curling for java agent is returning a 404

### ISSUE THREAD

**Issue by drewglass** [djoos-cookbooks/newrelic#266]
curl for the java agent is returning a 404 but I get a 200 in a browser.

```
$  curl -I  http://download.newrelic.com/newrelic/java-agent/newrelic-agent/current/
HTTP/1.1 404 Not Found
Last-Modified: Tue, 20 Sep 2016 22:13:40 GMT
x-amz-version-id: NLCnxJuorRTx5WOnoeySmdST9dAqQyv1
ETag: "a33b4d0f06bf8a8f61a6898e363bdbdb"
x-amz-error-code: NoSuchKey
x-amz-error-message: The specified key does not exist.
x-amz-error-detail-Key: newrelic/java-agent/newrelic-agent/current/index.html
x-amz-request-id: AE3A9E8C693EF743
x-amz-id-2: nrhCfQK6C74mH2f9pqlkw0PX2L54HrseJcvvVQd729sVlXQnKyRPKFoDKcW6iMplAXzdCrb+QDI=
Content-Type: text/html
Server: AmazonS3
Content-Length: 8129
Accept-Ranges: bytes
Date: Thu, 22 Sep 2016 18:44:17 GMT
Via: 1.1 varnish
Age: 2461
Connection: keep-alive
X-Served-By: cache-ord1728-ORD
X-Cache: HIT
X-Cache-Hits: 2
X-Timer: S1474569857.467308,VS0,VE0```

**Comment by drewglass** [comment]
Here is the line: https://github.com/djoos-cookbooks/newrelic/blob/master/providers/agent_java.rb#L46

It is an issue with New Relic.

**Comment by tjwallace** [comment]
They may have changed their download URLs? [See this page](https://docs.newrelic.com/docs/agents/manage-apm-agents/maintenance/agent-established-releases#java).

**Comment by mcmahoniel** [comment]
Hi folks. We (very) recently switched these downloads to S3. It looks like your script is attempting to download the index page, then parsing that page, and, finally, picking the most appropriate download link. Since there's no `index.html` available, you're getting stopped short (see the `curl` response above).

**Comment by pmacdougall** [comment]
Would be great if the download URL in the cookbook was an attribute so we could easily override it.  Not just to work around this but to also have a more localized location to download from.

That said it seems the current issue is because https://download.newrelic.com/newrelic/java-agent/newrelic-agent/current/ is actually responding with a 404, but most browsers will render the page so you can still see the file list.

> curl -I https://download.newrelic.com/newrelic/java-agent/newrelic-agent/current/
HTTP/1.1 404 Not Found
Last-Modified: Tue, 20 Sep 2016 22:13:40 GMT
x-amz-version-id: NLCnxJuorRTx5WOnoeySmdST9dAqQyv1
ETag: "a33b4d0f06bf8a8f61a6898e363bdbdb"
x-amz-error-code: NoSuchKey
x-amz-error-message: The specified key does not exist.
x-amz-error-detail-Key: newrelic/java-agent/newrelic-agent/current/index.html
x-amz-request-id: 0C6172A37DE4710D
x-amz-id-2: 9T9M6o1KgMPo/r8wrEpI031upusBq+WwNpBWgZ1LhxHy7OGI+/uCRhcx4G2Jz0dGCKAwDfPQOR4=
Content-Type: text/html
Server: AmazonS3
Content-Length: 8129
Accept-Ranges: bytes
Date: Thu, 22 Sep 2016 21:08:49 GMT
Via: 1.1 varnish
Age: 266
Connection: keep-alive
X-Served-By: cache-sea1926-SEA
X-Cache: HIT
X-Cache-Hits: 1
X-Timer: S1474578529.520986,VS0,VE0

**Comment by drewglass** [comment]
I resolved the issue by specifying the version of the Java agent.

**Comment by pmacdougall** [comment]
Thanks @drewglass . I was just about to report the same thing.

**Comment by joelvanvelden** [comment]
Thanks @drewglass 
As you say setting the version works wonders (the java_agent already had an attribute for this)

New relic are doing something very odd - the index page is now returning a 404 status and the page itself is all javascript just to render a index page. Not sure what they're trying to achieve.

Fixing the logic to find the latest version might not be too simple.

**Comment by spudnic** [comment]
I reached out to new relic directly to ask them about this and how they would like the community to pull the latest version of there agent.  We will see what they say.

**Comment by borislopezaraoz** [comment]
is this the right place to download from: https://download.newrelic.com/newrelic/java-agent/newrelic-agent/current/ ?
as @joelvanvelden said it is returning 404 and is now javascript generated.
and anyway is not good that the cookbook depends on parsing a web page.

I think @tjwallace is right, the page he linked links to "Java established releases" (from January)
https://oss.sonatype.org/content/repositories/releases/com/newrelic/agent/java/newrelic-java/3.25.0/
This is older than current from above link (3.32.0 September).

It is kind of confusing, I will also contact newrelic support about what is the best to automate the installation.

**Comment by borislopezaraoz** [comment]
From newrelic support:

> There was previously an issue with the downloads page where it would not render the download links properly outside of a browser. That issue has since been resolved so you can once again download the latest version of the agent via curl or wget by passing the URL directly.
For example:
$ curl -O https://download.newrelic.com/newrelic/java-agent/newrelic-agent/current/newrelic-java.zip
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100 8230k  100 8230k    0     0  14.5M      0 --:--:-- --:--:-- --:--:-- 14.5M
$ ls
newrelic-java.zip
If you'd like to use the Chef cookbook, the creators of the cookbook may need to update their logic for our updated downloads page. Alternatively, if you'd like to pass a URL directly, the URL, https://download.newrelic.com/newrelic/java-agent/newrelic-agent/current/newrelic-java.zip, will always point to our latest Java agent release.

**Comment by borislopezaraoz** [comment]
The problem as pointed out by @drewglass is here:
https://github.com/djoos-cookbooks/newrelic/blob/master/providers/agent_java.rb#L46

It is better to avoid parsing that page, just download directly: https://download.newrelic.com/newrelic/java-agent/newrelic-agent/#{version}/newrelic-java.zip
And then unzip only newrelic.jar from the zip. Something like this (not tested):

```ruby
def agent_jar
  zip_file = 'newrelic-java.zip'
  version = new_resource.version
  version = 'current' if version == 'latest'
  https_download = "https://download.newrelic.com/newrelic/java-agent/newrelic-agent/#{version}/#{zip_file}"

  cache_dir = Chef::Config[:file_cache_path]
  tmp_file = "#{cache_dir}/#{zip_file}"

  remote_file tmp_file do
    source https_download
    mode '0664'
    action :create
    notifies :run, "bash[unzip-#{tmp_file}]", :immediately
  end

  package 'unzip'

  bash "unzip-#{tmp_file}" do
    user new_resource.app_user
    group new_resource.app_group
    action :nothing
    code <<-EOH
      unzip -oj "#{tmp_file}" "newrelic/newrelic.jar" -d "#{new_resource.install_dir}"
    EOH
  end
end
```

**Comment by joelvanvelden** [comment]
Another possibility is peering directly into the new relic download archive s3 bucket directly to get a directory listing (albeit in xml) using a url like https://nr-downloads-main.s3.amazonaws.com/?list-type=2&delimiter=/&prefix=newrelic/java-agent/newrelic-agent/current/

Which would allow determining the latest version (eg "3.32.0"), then drop back to the current logic of requesting (guessing) the download url from the version (which still works). I doubt new relic would have a problem as the actual download itself would still be going via their CDN in-front of S3.

The upside of this method is the logic stays the same as it currently does, albeit we now have to parse xml (debatably a more accurate way of scraping than before).

Unfortunately I don't have time to work on this as my client has moved on -- preferring to set the version explicitly instead.

**Comment by Brunzer** [comment]
Just tested out @borislopezaraoz proposed solution and it worked. Nice simple solution that kept the flow very similar to the original code.

**Comment by mvdkleijn** [comment]
@djoos Any chance this issue can get some love? :smile: :sparkling_heart:

**Comment by mvdkleijn** [comment]
@drewglass how did you specify the version? I'm stuck at latest... probably I'm overlooking something simple. The cookbook docs say we need to set `https_download` but then I get the following:

`undefined method `https_download' for Custom resource newrelic_agent_java from cookbook newrelic`

**Comment by djoos** [comment]
Hi guys,

I'll try to free up some time next week to look into this issue - thanks!

David

**Comment by mvdkleijn** [comment]
Hi @djoos - Did you get a chance to look at this? The solution proposed by @borislopezaraoz seems fine to resolve this issue?

**Comment by cakeface** [comment]
I'd love for this to be fixed soon but in case anyone else gets hung up, specifying the version currently works around the issue.  Specifying the version may or may not be a good idea anyway as I don't like it when repeated runs of my chef scripts have different behavior but that's maybe a different story.

**Comment by djoos** [comment]
Hi all,

I've spotted @Brunzer's PR (#272) and got some minor (nitpicking) remarks. Once addressed, I'm happy to merge this in...

Thanks!

**Comment by djoos** [comment]
Merged and shipped in 2.23.0 together with CentOS 7.2 integration tests.

I'll close this issue for now - but don't hesitate to reach out if you have any further questions or comments!
